### DISCUSSION TITLE
Querybuilder for LATERAL queries

### DISCUSSION THREAD

**Discussion by valkum** [diesel-rs/diesel#3132] [Q&A]
We have the following problem:
We have A has many B.

We want to limit the amount of has_many entities B when we return an entity A.
To not run into the n+1 problem (calling select with LIMIT for each A) normally you would use belonging_to(&[A1,A2,...]).
But now we are not longer able to limit these correctly. 
What we need is basically a limit by group.

In Postgres this is doable by using a LATERAL sub-query, which should be better for this use case instead of using a window based row_number approach.

E.g.
From this query for the normal `belonging_to`: 
`SELECT e.* FROM "e" WHERE "e"."event_id" IN ($1, $2, $3)`
we need to end up with: 
`SELECT DISTINCT e_top.* FROM e, LATERAL (SELECT * FROM e AS e_inner WHERE e.rel_id = e_inner.rel_id LIMIT 2) e_top WHERE e.rel_id IN ($1, $2, $3);`

What would be the best way to construct this?
Ideally being able to call `E::belonging_to().limit_by_group(num, field)` would be nice, but that is not possible I guess.
For that to work limit_by_group would need to rewrite the select (which might be possible) but would also create a sub-query based on the current query returned by `belonging_to`.

Is there even a way to have a function to generate such a query based on an input query?

**Comment by weiznich** [comment]
The [Extending diesel](https://diesel.rs/guides/extending-diesel.html) guide has some similar examples that could be useful here. Generally speaking it should be possible to construct something like that, although I do not have a complete example that covers your exact use case.
